PR: Unify UI of editors in Variable Explorer and simplify code

jitseniesen opened the pull request:
<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes


* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

This PR is a follow-up to PR #21312 which added the Refresh button, in order to make the UI and the code in the array editor, collections editor, dataframe editor and object explorer more uniform, as suggested in issue #21556:
* Redistribute and reorder the actions in the context menus and the toolbars so that the UI is more logical and consistent.
* Move buttons in bottom left corner of dataframe editor (Format, Resize, Background Color, Column Min/Max) to toolbar.
* Tweak the UI for consistency: no vertical space between toolbar and main widget, and standardize vertical space between main widget and "Close" button at the bottom and align right edge.
* Use standard functions from SpyderWidgetMixin to create actions, menus and toolbars; this changes the appearance of the toolbar in the object explorer.
* Minor refactoring to simplify the code.

After the PR, the editors look as follows:

![followup-array-new3](https://github.com/spyder-ide/spyder/assets/7941918/8c9ae009-2939-49c1-892e-af51d981eee9)
![followup-coll-new3](https://github.com/spyder-ide/spyder/assets/7941918/90bcb810-dded-48a7-bcf2-b3f0b78945ce)
![followup-df-new3](https://github.com/spyder-ide/spyder/assets/7941918/069aea23-7e18-494b-be59-80b97432f83f)
![followup-obj-new2](https://github.com/spyder-ide/spyder/assets/7941918/4b45559e-2f34-43be-afc4-0a7ebacbdb55)
![followup-varexp-new3](https://github.com/spyder-ide/spyder/assets/7941918/e08f26b5-3f98-4729-80d8-edd3dff0ff8e)

Here are the corresponding images from before the PR:

![followup-array-old](https://github.com/spyder-ide/spyder/assets/7941918/a8030989-e4ad-41d8-8f01-507338581e80)
![followup-coll-old](https://github.com/spyder-ide/spyder/assets/7941918/bf195c45-0271-4c59-8e79-c11670d1cf4d)
![followup-df-old](https://github.com/spyder-ide/spyder/assets/7941918/51b3b959-cade-4b35-a9b0-4dac1d5ccbd4)
![followup-obj-old](https://github.com/spyder-ide/spyder/assets/7941918/c0c80962-01d5-457f-ac19-db3bfebb22cb)
![followup-varexp-old](https://github.com/spyder-ide/spyder/assets/7941918/1b5ec865-ee75-43bb-b73a-a822946512ab)


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes #21556


### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct: Jitse Niesen

<!--- Thanks for your help making Spyder better for everyone! --->

jitseniesen commented:
I added a commit which reflects the latest comment on the issue regarding the UI. I also updated the screenshots at the top of this pull request and added a screenshot of the Variable Editor itself.

CAM-Gerlach reviewed (commented):
Thanks so much for these great improvements, @jitseniesen ! (Didn't see this until now, sorry,  since I was watching the issue but not this PR)

I just have a few followup comments/suggestions:

* [x] Seems like `insert above` is _below_ `insert below` for some reason? Seems like they should be reversed (as in my original suggestion)
* [x] Seems like the refresh button is on the left rather than the right in the collections editor toolbar?
* [x] The (normally blank) circular progress indicator leaves a gap between the refresh button and the options menu; we should move that to the left of the right toolbar button group (i.e. to the left of the search button)
* [x] Upon seeing how they look in the UI, I think we should split the row and column actions in DF viewer context menu and toolbar into separate groups, instead of having them all in one big one. This would make them easier to navigate and not seem too overwhelming to users. So instead of `Insert above`, `Insert below`, `Insert before` `Insert after` `Duplicate row` `Duplicate column` `Remove Row` `Remove Column`; I suggest `Insert above`, `Insert below`, `Duplicate row` `Remove Row` `|`  `Insert before` `Insert after` `Duplicate column` `Remove Column`.
* [ ] Finally, I went back and forth on this a lot before and almost suggested it previously but didn't, but given the column resize buttons in the array and df editors affect only the presentation rather than the content itself, I think we should move them to immediately to the left of the refresh button instead of as a separate group at the beginning of the toolbar. This also puts core content-focused toolbar actions first in the df editor, as is generally expected and done in the other toolbars.

Thanks!

jitseniesen commented:
Thanks @CAM-Gerlach for your further comments. I implemented them and updated the screenshots at the top of this PR.

> Seems like `insert above` is _below_ `insert below` for some reason? Seems like they should be reversed (as in my original suggestion)

Done, under the assumption that you meant that `insert after` and `insert before` should be reversed.

> Seems like the refresh button is on the left rather than the right in the collections editor toolbar?

Done.

> The (normally blank) circular progress indicator leaves a gap between the refresh button and the options menu; we should move that to the left of the right toolbar button group (i.e. to the left of the search button)

Yes, I noticed that as well and I was planning to ask you about it, but I forgot, so I'm glad you picked that up. I did not find a clear way to do this using Spyder's API so I ended up overriding the private `_setup()` method of `PluginMainWidget`.

> Upon seeing how they look in the UI, I think we should split the row and column actions in DF viewer context menu and toolbar into separate groups, instead of having them all in one big one. This would make them easier to navigate and not seem too overwhelming to users. So instead of `Insert above`, `Insert below`, `Insert before` `Insert after` `Duplicate row` `Duplicate column` `Remove Row` `Remove Column`; I suggest `Insert above`, `Insert below`, `Duplicate row` `Remove Row` `|`  `Insert before` `Insert after` `Duplicate column` `Remove Column`.

Done. 

> Finally, I went back and forth on this a lot before and almost suggested it previously but didn't, but given the column resize buttons in the array and df editors affect only the presentation rather than the content itself, I think we should move them to immediately to the left of the refresh button instead of as a separate group at the beginning of the toolbar. This also puts core content-focused toolbar actions first in the df editor, as is generally expected and done in the other toolbars.

Done. This does mean that the toolbar in the array editor only has buttons on the right and nothing on the left, but so be it. I did the same move in the collection editor.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/dataframeeditor.py:
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
-            if item is not None:
-                if item.text() != 'Convert to':
-                    self.toolbar.addAction(item)
+        self.toolbar.addAction(self.dataTable.insert_action_above)
+        self.toolbar.addAction(self.dataTable.insert_action_below)
+        self.toolbar.addAction(self.dataTable.duplicate_row_action)
+        self.toolbar.addAction(self.dataTable.remove_row_action)
+        self.toolbar.addSeparator()
+        self.toolbar.addAction(self.dataTable.insert_action_before)
+        self.toolbar.addAction(self.dataTable.insert_action_after)
+        self.toolbar.addAction(self.dataTable.duplicate_col_action)
+        self.toolbar.addAction(self.dataTable.remove_col_action)
```
Is it not possible to use `add_item_to_toolbar` (which support sections) here?

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/dataframeeditor.py:
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
+        stretcher = self.create_stretcher('Toolbar stretcher')
+        self.toolbar.addWidget(stretcher)
+
+        self.toolbar.addAction(self.dataTable.resize_action)
+        self.toolbar.addAction(self.dataTable.resize_columns_action)
+        self.toolbar.addAction(self.refresh_action)
```
And finally you'd need to add this actions (I'm not sure if you need to declare a section for them or not).

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/dataframeeditor.py:
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
+        self.toolbar.addAction(self.refresh_action)
+
+        options_menu = self.create_menu('Options menu', register=False)
+        options_menu.add_action(self.bgcolor_action)
+        options_menu.add_action(self.bgcolor_global_action)
+        options_menu.add_action(self.format_action)
```
```suggestion
        for action in [self.bgcolor_action, self.bgcolor_global_action,
                       self.format_action]:
            self.add_item_to_menu(action, options_menu)

```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/dataframeeditor.py:
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
+            register=False
+        )
+        options_button.setPopupMode(QToolButton.InstantPopup)
+        options_button.setMenu(options_menu)
```
```suggestion
        options_button.setMenu(options_menu)

```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/dataframeeditor.py:
```diff
@@ -1774,10 +1808,36 @@ def set_data_and_check(self, data) -> bool:
+        )
+        options_button.setPopupMode(QToolButton.InstantPopup)
+        options_button.setMenu(options_menu)
+        self.toolbar.addWidget(options_button)
```
This should also use the toolbar API.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/dataframeeditor.py:
```diff
@@ -1790,14 +1850,15 @@ def save_and_close_enable(self, top_left, bottom_right):
         header_menu = [edit_header_action]
-        menu = SpyderMenu(self)
+        menu = self.create_menu('Context header menu', register=False)
         add_actions(menu, header_menu)
```
```suggestion
        self.add_item_to_menu(menu, header_menu)
```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/main_widget.py:
```diff
@@ -435,6 +436,26 @@ def setup(self):
+            if action.defaultWidget() == self.get_options_menu_button():
+                options_menu_action = action
+
+        corner_widget.insertAction(options_menu_action, self.search_action)
+        corner_widget.insertAction(options_menu_action, self.filter_button)
+        corner_widget.insertAction(options_menu_action, self.refresh_action)
```
```suggestion
        for action in [self.search_action, self.filter_button,
                       self.refresh_action]:
            corner_widget.insertAction(options_menu_action, action)
```

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:
```diff
@@ -15,26 +15,23 @@
-from spyder.utils.qthelpers import (
-    add_actions, create_toolbutton, qapplication, safe_disconnect)
-from spyder.utils.stylesheet import PANES_TOOLBAR_STYLESHEET
+from spyder.utils.qthelpers import add_actions, qapplication
```
```suggestion
from spyder.utils.qthelpers import qapplication
```

As before, `add_actions` shouldn't be necessary, so please replace its usage with the appropriate API method.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:
```diff
@@ -205,122 +179,125 @@ def set_value(self, obj):
+            option='show_callable_attributes',
+            register_action=False
         )
+        self.toolbar.add_item(self.toggle_show_callable_action)
```
I think it be better if you introduce all items after creating them first (i.e. at the end of this method) and by using `add_item_to_toolbar`.

ccordoba12 commented on spyder/plugins/variableexplorer/widgets/objectexplorer/objectexplorer.py:
```diff
@@ -205,122 +179,125 @@ def set_value(self, obj):
         self.central_splitter.addWidget(bottom_pane_widget)
 
         group_box = QGroupBox(_("Details"))
+        group_box.setStyleSheet('QGroupBox '
+                                '{margin-bottom: 0px; margin-right: -2px;}')
```
```suggestion
        group_box.setStyleSheet(
            'QGroupBox {margin-bottom: 0px; margin-right: -2px;}'
        )
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -45,24 +45,21 @@
 from spyder.utils.misc import getcwd_or_home
-from spyder.utils.qthelpers import (
-    add_actions, create_action, MENU_SEPARATOR, mimedata2url)
+from spyder.utils.qthelpers import add_actions, MENU_SEPARATOR, mimedata2url
```
```suggestion
from spyder.utils.qthelpers import mimedata2url
```

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -644,99 +642,146 @@ def setup_table(self):
+            self.imshow_action
         ]
-        add_actions(menu, self.menu_actions)
+        add_actions(menu, menu_actions)
```
This menu should also be built as explained above, i.e. by declaring sections and then using `add_item_to_menu`.

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -644,99 +642,146 @@ def setup_table(self):
 
-        self.empty_ws_menu = SpyderMenu(self)
+        self.empty_ws_menu = self.create_menu('Empty ws', register=False)
         add_actions(
```
Please replace this usage of `add_actions` (I can't leave a suggestion for it).

ccordoba12 commented on spyder/widgets/collectionseditor.py:
```diff
@@ -1464,27 +1509,38 @@ def __init__(self, parent, data, namespacebrowser=None,
+            register_action=None
         )
+
+        toolbar = self.create_toolbar('Editor toolbar', register=False)
```
If possible, please build this toolbar with our API.

pep8speaks commented:
Hello @jitseniesen! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`spyder/plugins/variableexplorer/widgets/dataframeeditor.py`](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/plugins/variableexplorer/widgets/dataframeeditor.py):

> [Line 70:1](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/plugins/variableexplorer/widgets/dataframeeditor.py#L70): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 149:1](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/plugins/variableexplorer/widgets/dataframeeditor.py#L149): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 151:1](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/plugins/variableexplorer/widgets/dataframeeditor.py#L151): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 182:1](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/plugins/variableexplorer/widgets/dataframeeditor.py#L182): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '
> [Line 184:1](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/plugins/variableexplorer/widgets/dataframeeditor.py#L184): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '











* In the file [`spyder/widgets/collectionseditor.py`](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/widgets/collectionseditor.py):

> [Line 827:9](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/widgets/collectionseditor.py#L827): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented
> [Line 835:9](https://github.com/spyder-ide/spyder/blob/4e2c5598b5495091996ea7522b492729cdd488c4/spyder/widgets/collectionseditor.py#L835): [E122](https://duckduckgo.com/?q=pep8%20E122) continuation line missing indentation or outdented

jitseniesen commented:
@ccordoba12 As requested, the PR is now using add_item_to_menu() and add_item_to_toolbar() from SpyderWidgetMixin to build the menus and toolbars, and it's using sections to add the separators. For consistency, I also "declared" all the names of actions and widgets in enum-style classes like DataframeEditorActions. It does make the code more verbose but I guess that is the price to pay for consistency. I also addressed your other comments.

The PR does need a rebase, but I don't want to mess up your review process. So please let me know when you want to rebase.

CAM-Gerlach reviewed (approved):
LGTM now from my side; thanks @jitseniesen !

ccordoba12 reviewed (approved):
This is a terrific improvement!! Thanks @jitseniesen for your hard work on this!

ccordoba12 commented:
Note: There are some really minor issues about this that I'd like to discuss with @jitseniesen and @CAM-Gerlach, but I'm going to merge because @dalthviz needs this so he can finish the migration of the editor to the new API.

So, I'll open a new issue about them.
